third/nmh/man/send.man
.\"
.\" %nmhwarning%
.\" $Id: send.man,v 1.1.1.1 1999-02-07 18:14:22 danw Exp $
.\"
.\" include the -mh macro file
.so %etcdir%/tmac.h
.\"
.TH SEND %manext1% MH.6.8 [%nmhversion%]
.SH NAME
send \- send a message
.SH SYNOPSIS
.in +.5i
.ti -.5i
send
\%[\-alias\ aliasfile]
\%[\-draft]
\%[\-draftfolder\ +folder]
.br
\%[\-draftmessage\ msg] \%[\-nodraftfolder]
.br
\%[\-filter\ filterfile] \%[\-nofilter]
\%[\-format] \%[\-noformat]
.br
\%[\-forward] \%[\-noforward]
\%[\-mime] \%[\-nomime]
\%[\-msgid]
.br
\%[\-nomsgid]
\%[\-push] \%[\-nopush]
\%[\-split\ seconds]
.br
\%[\-verbose] \%[\-noverbose]
\%[\-watch] \%[\-nowatch]
.br
\%[\-width\ columns]
\%[file\ ...] 
\%[\-version]
\%[\-help]
.in -.5i
.SH DESCRIPTION
\fISend\fR will cause each of the specified files to be delivered
to each of the destinations in the \*(lqTo:\*(rq, \*(lqcc:\*(rq,
\*(lqBcc:\*(rq, and \*(lqFcc:\*(rq fields of the message.  If \fIsend\fR
is re\-distributing a message, as invoked from \fIdist\fR, then the
corresponding \*(lqResent\-xxx\*(rq fields are examined instead.

By default, \fIsend\fR uses the program \fIpost\fR(8) to do the actual
delivery of the messages, although this can be changed by defining the
\fIpostproc\fR profile component.  Most of the features attributed to
\fIsend\fR are actually performed by \fIpost\fR.

If `\-push' is specified, \fIsend\fR will detach itself from the user's
terminal and perform its actions in the background.  If \fIpush\fR\0'd
and the draft can't be sent, then an error message will be sent (using
the mailproc) back to the user.  If `\-forward' is given, then a copy
of the draft will be attached to this failure notice.  Using `\-push'
differs from putting \fIsend\fR in the background because the output is
trapped and analyzed by \fInmh\fR.

If `\-verbose' is specified, \fIsend\fR will indicate the interactions
occurring with the transport system, prior to actual delivery.
If `\-watch' is specified \fIsend\fR will monitor the delivery of local
and network mail.  Hence, by specifying both switches, a large detail
of information can be gathered about each step of the message's entry
into the transport system.

The `\-draftfolder\ +folder' and `\-draftmessage\ msg' switches invoke
the \fInmh\fR draft folder facility.  This is an advanced (and highly
useful) feature.  Consult the \fImh-draft\fR(5) man page for more
information.

If `\-split' is specified, \fIsend\fR will split the draft into one
or more partial messages prior to sending.  This makes use of the
MIME features in nmh.  Note however that if \fIsend\fR is
invoked under \fIdist\fR\0(1), then this switch is ignored\0--\0it makes
no sense to redistribute a message in this fashion.  Sometimes you want
\fIsend\fR to pause after posting a partial message.  This is usually
the case when you are running \fIsendmail\fR and expect to generate a
lot of partial messages.  The argument to `\-split' tells it how long
to pause between postings.

\fISend\fR with no \fIfile\fR argument will query whether the draft
is the intended file, whereas `\-draft' will suppress this question.
Once the transport system has successfully accepted custody of the
message, the file will be renamed with a leading comma, which allows
it to be retrieved until the next draft message is sent.  If there are
errors in the formatting of the message, \fIsend\fR will abort with a
(hopefully) helpful error message.

If a \*(lqBcc:\*(rq field is encountered, its addresses will be used for
delivery, and the \*(lqBcc:\*(rq field will be removed from the message
sent to sighted recipients.  The blind recipients will receive an entirely
new message with a minimal set of headers.  Included in the body of the
message will be a copy of the message sent to the sighted recipients.
If `\-filter\ filterfile' is specified, then this copy is filtered
(re\-formatted) by \fImhl\fR prior to being sent to the blind recipients.
Alternately, if you specify the `-mime' switch, then \fIsend\fR will
use the MIME rules for encapsulation.

Prior to sending the message, the fields \*(lqFrom:\ user@local\*(rq,
and \*(lqDate:\ now\*(rq will be appended to the headers in the message.
If the environment variable \fB$SIGNATURE\fR is set, then its value
is used as your personal name when constructing the \*(lqFrom:\*(rq
line of the message.  If this environment variable is not set, then
\fIsend\fR will consult the profile entry \*(lqSignature\*(rq for
this information.  On hosts where \fInmh\fR was configured with the UCI
option, if \fB$SIGNATURE\fR is not set and the \*(lqSignature\*(rq profile
entry is not present, then the file \fB$HOME\fR/.signature is consulted.
If `\-msgid' is specified, then a \*(lqMessage\-ID:\*(rq field will also
be added to the message.

If \fIsend\fR is re\-distributing a message (when invoked by
\fIdist\fR\0), then \*(lqResent\-\*(rq will be prepended to each of these
fields: \*(lqFrom:\*(rq, \*(lqDate:\*(rq, and \*(lqMessage\-ID:\*(rq.
If the message already contains a \*(lqFrom:\*(rq field, then a
\*(lqSender: user@local\*(rq field will be added as well.  (An already
existing \*(lqSender:\*(rq field is an error!)

By using the `\-format' switch, each of the entries in the \*(lqTo:\*(rq
and \*(lqcc:\*(rq fields will be replaced with \*(lqstandard\*(rq
format entries.  This standard format is designed to be usable by all
of the message handlers on the various systems around the Internet.
If `\-noformat' is given, then headers are output exactly as they appear
in the message draft.

If an \*(lqFcc:\ folder\*(rq is encountered, the message will be copied
to the specified folder for the sender in the format in which it will
appear to any non\-Bcc receivers of the message.  That is, it will have
the appended fields and field reformatting.  The \*(lqFcc:\*(rq fields
will be removed from all outgoing copies of the message.

By using the `\-width\ columns' switch, the user can direct \fIsend\fR
as to how long it should make header lines containing addresses.

The files specified by the profile entry \*(lqAliasfile:\*(rq and any
additional alias files given by the `\-alias aliasfile' switch will be
read (more than one file, each preceded by `\-alias', can be named).
See \fImh\-alias\fR\0(5) for more information.
.Fi
^$HOME/\&.mh\(ruprofile~^The user profile
.Pr
^Path:~^To determine the user's nmh directory
.Ps
^Draft\-Folder:~^To find the default draft\-folder
.Ps
^Aliasfile:~^For a default alias file
.Ps
^Signature:~^To determine the user's mail signature
.Ps
^mailproc:~^Program to post failure notices
.Ps
^postproc:~^Program to post the message
.Sa
comp(1), dist(1), forw(1), repl(1), mh\-alias(5), post(8)
.De
`file' defaults to <mh\-dir>/draft
.Ds
`\-alias %etcdir%/MailAliases'
.Ds
`\-nodraftfolder'
.Ds
`\-nofilter'
.Ds
`\-format'
.Ds
`\-forward'
.Ds
`\-nomime'
.Ds
`\-nomsgid'
.Ds
`\-nopush'
.Ds
`\-noverbose'
.Ds
`\-nowatch'
.Ds
`\-width\ 72'
.Co
None
.Bu
Under some configurations, it is not possible to monitor the mail delivery
transaction; `\-watch' is a no-op on those systems.
.sp
Using `\-split\00' doesn't work correctly.
.En


doc/news
This file is intended to keep track of user-visible changes to the
release, for the construction of the Athena release notes.

Changes from release 8.0 to release 8.1:

	* Athena version numbers are now dotted triplets, e.g. 8.1.0
	  or 8.1.12.  Paths which include all or part of the Athena
	  version number now take the form "pathname-version",
	  e.g. "srvd-8.1" or "dash.menus-8.1"

	* In release 8.0, running "xmkmf" with the default path would
	  generate a Makefile which looked for its include files and
	  libraries in /mit/x11, which could be one of several
	  lockers.  Now it is preferrable to use the vendor X
	  libraries, so this bit of glue is going away.  There will be
	  no /usr/athena/bin/xmkmf any more, nor any symlinks in
	  /usr/athena pointing to the x11 locker.

	  Users can still do something like "add -f x11r5" and then
	  "xmkmf", and things should work, if the x11r5 locker is
	  built for their platform.

	* Athena bitmaps have moved to /usr/athena/include/X11/bitmaps.
	  Previously, they were in /usr/athena/lib/X11/bitmaps before
	  because of the /mit/x11 glue.

	* The symlink /bin/csh -> /bin/athena/tcsh has gone away.  The
	  update to 8.0 will fix /bin/csh references in /etc/passwd
	  and /etc/passwd.local to refer to /bin/athena/tcsh instead.

	* emacs18 has gone away.

	* AFS includes and libraries are installed in
	  /usr/afsws/include and /usr/afsws/lib.

	* wcl includes and libraries are installed in /usr/athena.

	* The basic krb5 headers and libraries are installed in
	  /usr/athena.

	* emacs has been upgraded from 19.30 to 19.34.  There should
	  be no significant user-visible changes.

	* RCS has been upgraded from version 4 to version 5.7.

	* x3270 has been upgraded from version 1.2 to version 3.1.1.3.

	* sysinfo is now available under Irix.

	* /bin/athena/tcsh now has Hesiod support enabled, which means
	  that "~lockername" will work in a pathname in most cases.
	  (It won't work for users with local passwd information who
	  don't have their lockers attached.)

	* For improved ease of debugging, many binaries are installed
	  with symbol tables, and compiled without optimization.
	  Therefore, the contents of /bin/athena and /usr/athena/bin
	  may be somewhat larger, and the binaries may consume more
	  CPU cycles, although it's unlikely that anyone will notice
	  the speed difference.


third/nmh/uip/send.c

/*
 * send.c -- send a composed message
 *
 * $Id: send.c,v 1.1.1.1 1999-02-07 18:14:16 danw Exp $
 */

#include <h/mh.h>
#include <fcntl.h>
#include <errno.h>
#include <signal.h>


static struct swit switches[] = {
#define	ALIASW                 0
    { "alias aliasfile", 0 },
#define	DEBUGSW                1
    { "debug", -5 },
#define	DRAFTSW                2
    { "draft", 0 },
#define	DFOLDSW                3
    { "draftfolder +folder", 6 },
#define	DMSGSW                 4
    { "draftmessage msg", 6 },
#define	NDFLDSW                5
    { "nodraftfolder", 0 },
#define	FILTSW                 6
    { "filter filterfile", 0 },
#define	NFILTSW                7
    { "nofilter", 0 },
#define	FRMTSW                 8
    { "format", 0 },
#define	NFRMTSW                9
    { "noformat", 0 },
#define	FORWSW                10
    { "forward", 0 },
#define	NFORWSW               11
    { "noforward", 0 },
#define MIMESW                12
    { "mime", 0 },
#define NMIMESW               13
    { "nomime", 0 },
#define	MSGDSW                14
    { "msgid", 0 },
#define	NMSGDSW               15
    { "nomsgid", 0 },
#define	PUSHSW                16
    { "push", 0 },
#define	NPUSHSW               17
    { "nopush", 0 },
#define	SPLITSW               18
    { "split seconds", 0 },
#define	UNIQSW                19
    { "unique", -6 },
#define	NUNIQSW               20
    { "nounique", -8 },
#define	VERBSW                21
    { "verbose", 0 },
#define	NVERBSW               22
    { "noverbose", 0 },
#define	WATCSW                23
    { "watch", 0 },
#define	NWATCSW               24
    { "nowatch", 0 },
#define	WIDTHSW               25
    { "width columns", 0 },
#define VERSIONSW             26
    { "version", 0 },
#define	HELPSW                27
    { "help", 4 },
#define BITSTUFFSW            28
    { "dashstuffing", -12 },
#define NBITSTUFFSW           29
    { "nodashstuffing", -14 },
#define	MAILSW                30
    { "mail", -4 },
#define	SAMLSW                31
    { "saml", -4 },
#define	SENDSW                32
    { "send", -4 },
#define	SOMLSW                33
    { "soml", -4 },
#define	CLIESW                34
    { "client host", -6 },
#define	SERVSW                35
    { "server host", -6 },
#define	SNOOPSW               36
    { "snoop", -5 },
    { NULL, 0 }
};

static struct swit anyl[] = {
#define	NOSW     0
    { "no", 0 },
#define	YESW     1
    { "yes", 0 },
#define	LISTDSW  2
    { "list", 0 },
    { NULL, 0 }
};

extern int debugsw;		/* from sendsbr.c */
extern int forwsw;
extern int inplace;
extern int pushsw;
extern int splitsw;
extern int unique;
extern int verbsw;

extern char *altmsg;		/*  .. */
extern char *annotext;
extern char *distfile;

extern int   errno;

int
main (int argc, char **argv)
{
    int msgp = 0, distsw = 0, vecp = 1;
    int isdf = 0, mime = 0;
    int msgnum, status;
    char *cp, *dfolder = NULL, *maildir = NULL;
    char buf[BUFSIZ], **ap, **argp, **arguments;
    char *msgs[MAXARGS], *vec[MAXARGS];
    struct msgs *mp;
    struct stat st;
#ifdef UCI
    FILE *fp;
#endif /* UCI */

#ifdef LOCALE
    setlocale(LC_ALL, "");
#endif
    invo_name = r1bindex (argv[0], '/');

    /* read user profile/context */
    context_read();

    arguments = getarguments (invo_name, argc, argv, 1);
    argp = arguments;

    vec[vecp++] = "-library";
    vec[vecp++] = getcpy (m_maildir (""));

    while ((cp = *argp++)) {
	if (*cp == '-') {
	    switch (smatch (++cp, switches)) {
		case AMBIGSW: 
		    ambigsw (cp, switches);
		    done (1);
		case UNKWNSW: 
		    adios (NULL, "-%s unknown\n", cp);

		case HELPSW: 
		    snprintf (buf, sizeof(buf), "%s [file] [switches]", invo_name);
		    print_help (buf, switches, 1);
		    done (1);
		case VERSIONSW:
		    print_version(invo_name);
		    done (1);

		case DRAFTSW: 
		    msgs[msgp++] = draft;
		    continue;

		case DFOLDSW: 
		    if (dfolder)
			adios (NULL, "only one draft folder at a time!");
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    dfolder = path (*cp == '+' || *cp == '@' ? cp + 1 : cp,
			    *cp != '@' ? TFOLDER : TSUBCWF);
		    continue;
		case DMSGSW: 
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    msgs[msgp++] = cp;
		    continue;
		case NDFLDSW: 
		    dfolder = NULL;
		    isdf = NOTOK;
		    continue;

		case PUSHSW: 
		    pushsw++;
		    continue;
		case NPUSHSW: 
		    pushsw = 0;
		    continue;

		case SPLITSW: 
		    if (!(cp = *argp++) || sscanf (cp, "%d", &splitsw) != 1)
			adios (NULL, "missing argument to %s", argp[-2]);
		    continue;

		case UNIQSW: 
		    unique++;
		    continue;
		case NUNIQSW: 
		    unique = 0;
		    continue;

		case FORWSW:
		    forwsw++;
		    continue;
		case NFORWSW:
		    forwsw = 0;
		    continue;

		case VERBSW: 
		    verbsw++;
		    vec[vecp++] = --cp;
		    continue;
		case NVERBSW:
		    verbsw = 0;
		    vec[vecp++] = --cp;
		    continue;

		case MIMESW:
		    mime++;
		    vec[vecp++] = --cp;
		    continue;
		case NMIMESW:
		    mime = 0;
		    vec[vecp++] = --cp;
		    continue;

		case DEBUGSW: 
		    debugsw++;	/* fall */
		case NFILTSW: 
		case FRMTSW: 
		case NFRMTSW: 
		case BITSTUFFSW:
		case NBITSTUFFSW:
		case MSGDSW: 
		case NMSGDSW: 
		case WATCSW: 
		case NWATCSW: 
		case MAILSW: 
		case SAMLSW: 
		case SENDSW: 
		case SOMLSW: 
		case SNOOPSW: 
		    vec[vecp++] = --cp;
		    continue;

		case ALIASW: 
		case FILTSW: 
		case WIDTHSW: 
		case CLIESW: 
		case SERVSW: 
		    vec[vecp++] = --cp;
		    if (!(cp = *argp++) || *cp == '-')
			adios (NULL, "missing argument to %s", argp[-2]);
		    vec[vecp++] = cp;
		    continue;
	    }
	} else {
	    msgs[msgp++] = cp;
	}
    }

    /*
     * check for "Aliasfile:" profile entry
     */
    if ((cp = context_find ("Aliasfile"))) {
	char *dp = NULL;

	for (ap = brkstring(dp = getcpy(cp), " ", "\n"); ap && *ap; ap++) {
	    vec[vecp++] = "-alias";
	    vec[vecp++] = *ap;
	}
    }

    if (dfolder == NULL) {
	if (msgp == 0) {
#ifdef WHATNOW
	    if ((cp = getenv ("mhdraft")) && *cp) {
		msgs[msgp++] = cp;
		goto go_to_it;
	    }
#endif /* WHATNOW */
	    msgs[msgp++] = getcpy (m_draft (NULL, NULL, 1, &isdf));
	    if (stat (msgs[0], &st) == NOTOK)
		adios (msgs[0], "unable to stat draft file");
	    cp = concat ("Use \"", msgs[0], "\"? ", NULL);
	    for (status = LISTDSW; status != YESW;) {
		if (!(argp = getans (cp, anyl)))
		    done (1);
		switch (status = smatch (*argp, anyl)) {
		    case NOSW: 
			done (0);
		    case YESW: 
			break;
		    case LISTDSW: 
			showfile (++argp, msgs[0]);
			break;
		    default:
			advise (NULL, "say what?");
			break;
		}
	    }
	} else {
	    for (msgnum = 0; msgnum < msgp; msgnum++)
		msgs[msgnum] = getcpy (m_maildir (msgs[msgnum]));
	}
    } else {
	if (!context_find ("path"))
	    free (path ("./", TFOLDER));

	if (!msgp)
	    msgs[msgp++] = "cur";
	maildir = m_maildir (dfolder);

	if (chdir (maildir) == NOTOK)
	    adios (maildir, "unable to change directory to");

	/* read folder and create message structure */
	if (!(mp = folder_read (dfolder)))
	    adios (NULL, "unable to read folder %s", dfolder);

	/* check for empty folder */
	if (mp->nummsg == 0)
	    adios (NULL, "no messages in %s", dfolder);

	/* parse all the message ranges/sequences and set SELECTED */
	for (msgnum = 0; msgnum < msgp; msgnum++)
	    if (!m_convert (mp, msgs[msgnum]))
		done (1);
	seq_setprev (mp);	/* set the previous-sequence */

	for (msgp = 0, msgnum = mp->lowsel; msgnum <= mp->hghsel; msgnum++) {
	    if (is_selected (mp, msgnum)) {
		msgs[msgp++] = getcpy (m_name (msgnum));
		unset_exists (mp, msgnum);
	    }
	}

	mp->msgflags |= SEQMOD;
	seq_save (mp);
    }

#ifdef WHATNOW
go_to_it:
#endif /* WHATNOW */

    if ((cp = getenv ("SIGNATURE")) == NULL || *cp == 0)
	if ((cp = context_find ("signature")) && *cp)
	    m_putenv ("SIGNATURE", cp);
#ifdef UCI
	else {
	    snprintf (buf, sizeof(buf), "%s/.signature", mypath);
	    if ((fp = fopen (buf, "r")) != NULL
		&& fgets (buf, sizeof buf, fp) != NULL) {
		    fclose (fp);
		    if (cp = strchr (buf, '\n'))
			*cp = 0;
		    m_putenv ("SIGNATURE", buf);
	    }
	}
#endif /* UCI */

    for (msgnum = 0; msgnum < msgp; msgnum++)
	if (stat (msgs[msgnum], &st) == NOTOK)
	    adios (msgs[msgnum], "unable to stat draft file");

    if ((annotext = getenv ("mhannotate")) == NULL || *annotext == 0)
	annotext = NULL;
    if (annotext && ((cp = getenv ("mhinplace")) != NULL && *cp != 0))
	inplace = atoi (cp);
    if ((altmsg = getenv ("mhaltmsg")) == NULL || *altmsg == 0)
	altmsg = NULL;	/* used by dist interface - see below */

    if ((cp = getenv ("mhdist"))
	    && *cp
	    && (distsw = atoi (cp))
	    && altmsg) {
	vec[vecp++] = "-dist";
	distfile = getcpy (m_scratch (altmsg, invo_name));
	if (link (altmsg, distfile) == NOTOK) {
	    if (errno != EXDEV
#ifdef EISREMOTE
		    && errno != EISREMOTE
#endif /* EISREMOTE */
		)
		adios (distfile, "unable to link %s to", altmsg);
	    free (distfile);
	    distfile = getcpy (m_tmpfil (invo_name));
	    {
		int in, out;
		struct stat st;

		if ((in = open (altmsg, O_RDONLY)) == NOTOK)
		    adios (altmsg, "unable to open");
		fstat(in, &st);
		if ((out = creat (distfile, (int) st.st_mode & 0777)) == NOTOK)
		    adios (distfile, "unable to write");
		cpydata (in, out, altmsg, distfile);
		close (in);
		close (out);
	    }	
	}
    } else {
	distfile = NULL;
    }

    if (altmsg == NULL || stat (altmsg, &st) == NOTOK) {
	st.st_mtime = 0;
	st.st_dev = 0;
	st.st_ino = 0;
    }
    if (pushsw)
	push ();

    status = 0;
    vec[0] = r1bindex (postproc, '/');
    closefds (3);

    for (msgnum = 0; msgnum < msgp; msgnum++) {
	switch (sendsbr (vec, vecp, msgs[msgnum], &st, 1)) {
	    case DONE: 
		done (++status);
	    case NOTOK: 
		status++;	/* fall */
	    case OK:
		break;
	}
    }

    context_save ();	/* save the context file */
    done (status);
}
